src/main/java/apoc/meta/tablesforlabels/PropertyTracker.java

```diff
@@ -0,0 +1,92 @@
+package apoc.meta.tablesforlabels;
+import org.neo4j.graphdb.Label;
+
+import java.util.*;
+
+public class PropertyTracker {
+    public String name;
+    public Set<String> types;
+    public List<String> sourceNodeLabels;
+    public List<String> targetNodeLabels;
+    public boolean mandatory;
+    public long observations;
+    public long nulls;
+
+    public PropertyTracker(String name) {
+        this.name = name;
+        sourceNodeLabels = new ArrayList<String>();
+        targetNodeLabels = new ArrayList<String>();
+        types = new HashSet<>(3);
+        mandatory = true;
+        observations = 0L;
+        nulls = 0L;
+    }
+
+    public void addObservation(Object value) {
+        observations++;
+
+        if (value == null) { nulls++; }
+
+        this.mandatory = this.mandatory && (value != null);
+        types.add(assignTypeName(value));
+    }
+
+    private String assignTypeName(Object value) {
+        String typeName = value.getClass().getCanonicalName();
+        if (typeMappings.containsKey(typeName)) {
+            return typeMappings.get(typeName);
+        }
+
+        return typeName.replace("java.lang.", "");
+    }
+
+    public List<String> propertyTypes() {
+        List<String> ret = new ArrayList<>(types);
+        Collections.sort(ret);
+        return ret;
+    }
+
+    /** Call when completed with the total number of nodes observed */
+    public void finished(long totalObservations) {
+        // As a final step, a given property can be missing on some nodes.
+        // If that occurred it wasn't tracked, and we need to mark it as optional
+        // because while it may always be populated *when its present* it can be missing
+        // from some nodes.
+        if (totalObservations > observations) {
+            mandatory = false;
+        }
+    }
+
+    public static final Map<String,String> typeMappings = new HashMap<String,String>();
+    static {
+        typeMappings.put("java.lang.String", "String");
```

Comment:
Prior to final approval, Shash please go through the types and ensure that we haven't missed anything there, or ideally make sure unit test coverage covers every possible type.

Some type documentation (or the entry point to it) is here: https://neo4j.com/docs/cypher-manual/current/functions/temporal/

The original code was written under 3.5.   I'm about 80% sure that in Neo4j 4.0 we didn't introduce any new types, but we need to make sure we have total coverage.   For each type we also need to cover the corresponding array type.

If we miss one, then the code will report the type as string.  The driver will similarly coerce it to string, and we'll end up with a bug report so we need to be sure we're 100% complete.  If you've done the review or are satisfied the unit tests cover this, then just flag a follow-up comment saying you looked into it and this point is good to go.

Reply:
As far as I can tell all types are accounted for as follows:

String
StringArray
Double
DoubleArray
Integer
IntegerArray
Long
LongArray
Point
PointArray
DateTime
DateTimeArray
Boolean
BooleanArray
Date
DateArray
LocalDateTime
LocalDateTimeArray
LocalTime
LocalTimeArray
Duration
DurationArray
Time
TimeArray